modules/hal_ethos_u/init.c

```diff
@@ -0,0 +1,164 @@
+
+	status = k_sem_init(sem, 1, 100);
+	if (status != 0) {
+		LOG_ERR("Failed to create semaphore with error - %d\n", status);
+	}
+	return (void *)sem;
+}
+
+void ethosu_semaphore_take(void *sem)
+{
+	int status = 0;
+
+	status = k_sem_take((struct k_sem *)sem, K_FOREVER);
+	if (status != 0) {
+		LOG_ERR("Failed to take semaphore with error - %d\n", status);
+	}
+}
+
+void ethosu_semaphore_give(void *sem)
+{
+	k_sem_give((struct k_sem *)sem);
+}
+
+struct ethosu_dts_info {
+	uint32_t base_addr;
+	uint32_t secure_enable;
+	uint32_t privilege_enable;
+	uint32_t irq;
+	uint32_t irq_priority;
+	uint32_t inst;
+};
+
+struct ethosu_data {
+	struct ethosu_driver drv;
+	void (*irq_config)(void);
+};
+
+static int ethosu_zephyr_init(const struct device *dev)
+{
+	const struct ethosu_dts_info *config = dev->config;
+	const struct ethosu_data *data = dev->data;
+	struct ethosu_driver *drv = (struct ethosu_driver *)&data->drv;
+	struct ethosu_driver_version version;
+
+	LOG_DBG("Ethos-U DTS info. base_address=0x%x, inst=%u, secure_enable=%u, privilege_enable=%u, irq=%u, "
+		"irq_priority=%u\n",
+		config->base_addr, config->inst, config->secure_enable, config->privilege_enable,
+		config->irq, config->irq_priority);
+
+	ethosu_get_driver_version(&version);
+	LOG_DBG("Version. major=%u, minor=%u, patch=%u\n", version.major, version.minor,
+		version.patch);
+	if (ethosu_init(drv, (void *)(config->base_addr), NULL, 0, config->secure_enable,
+			config->privilege_enable)) {
+		LOG_ERR("Failed to initialize NPU with ethosu_init().\n");
+		return -EINVAL;
+	}
+
+	data->irq_config();
+
+	return 0;
+}
+
+#define ETHOSU_DEVICE_INIT(n)									   \
+	static void ethosu_zephyr_irq_config_##n(void);						   \
+												   \
+	static struct ethosu_data ethosu_data_##n = { .irq_config =				   \
+							      &ethosu_zephyr_irq_config_##n };	   \
+												   \
+	static void ethosu_zephyr_irq_handler_##n(void)						   \
+	{											   \
+		struct ethosu_driver *drv = &ethosu_data_##n.drv;				   \
+		ethosu_irq_handler(drv);							   \
+	}											   \
+												   \
+	static void ethosu_zephyr_irq_config_##n(void)						   \
+	{											   \
+		IRQ_DIRECT_CONNECT(DT_INST_IRQN(n), DT_INST_IRQ(n, priority),			   \
+				   ethosu_zephyr_irq_handler_##n, 0);				   \
+		irq_enable(DT_INST_IRQN(n));							   \
+	}											   \
+												   \
+	static const struct ethosu_dts_info ethosu_dts_info_##n = {				   \
+		.base_addr = DT_INST_REG_ADDR(n),						   \
+		.secure_enable = DT_INST_PROP(n, secure_enable),				   \
+		.privilege_enable = DT_INST_PROP(n, privilege_enable),				   \
+		.irq = DT_INST_IRQN(n),								   \
+		.irq_priority = DT_INST_IRQ(n, priority),					   \
+		.inst = n,									   \
+	};											   \
+												   \
+	DEVICE_DT_INST_DEFINE(n, ethosu_zephyr_init, NULL, &ethosu_data_##n, &ethosu_dts_info_##n, \
+			      POST_KERNEL, CONFIG_KERNEL_INIT_PRIORITY_DEFAULT, NULL);
+
+DT_INST_FOREACH_STATUS_OKAY(ETHOSU_DEVICE_INIT);
```

Comment:
What is going on here? Why is this file instantiating devices with NULL API pointers?

Reply:
Thanks for taking time to review this code. This is the team's first contribution to Zephyr, so we are thankful for all the help we can get. And please bear with us if there are concepts in Zephyr that we have missed.

I didn't write this particular code myself, but it might be that we didn't know about the APIs when we implemented the device initialization.

Looking at the APIs under `zephyr/include/zephyr/drivers`, at first glance no API seems to be a perfect match Ethos-U. Ethos-U is a very specific piece of hardware that takes a command stream and up to 8 base pointers to compute a kernel in a neural network. If we have to implement a driver API, then I suspect we will need to add a new API just for Ethos-U.

The current implementation is very much based on how the Tensorflow Lite for Microcontrollers (TFLM) framework is designed. A neural network consists of a graph of kernels. The TFLM framework iterates of the nodes in the network and applies a kernel to each node. To accelerate a neural network on Ethos-U, the network model is run through the Vela compiler that replaces software kernels with a custom Ethos-U kernel. The implementation of the custom kernel can be found here. 

https://github.com/zephyrproject-rtos/tflite-micro/blob/main/tensorflow/lite/micro/kernels/ethos_u/ethosu.cc

The purpose of `init.c` is to register and enable the IRQ service route, and to initialize Ethos-U NPU driver based on the device tree entries.

Reply:
> Looking at the APIs under zephyr/include/zephyr/drivers, at first glance no API seems to be a perfect match Ethos-U. Ethos-U is a very specific piece of hardware that takes a command stream and up to 8 base pointers to compute a kernel in a neural network. If we have to implement a driver API, then I suspect we will need to add a new API just for Ethos-U.

OK. Then this feels like a misc driver to me (see drivers/misc/grove_lcd_rgb and include/zephyr/drivers/misc/grove_lcd/grove_lcd.h). @gmarull would you agree?

Reply:
The NPU driver code is placed in a separate module, because it is mirroring an parent repository. It is only the init code that is stored in the main Zephyr repository.
https://github.com/zephyrproject-rtos/hal_ethos_u

Do you mean that the code should be moved from `modules/hal_ethos_u/` to `drivers/misc/`?

Reply:
Yes, I mean that if it is a driver in the sense of a zephyr device driver that allocates `struct device` objects, as it is currently structured, it should be in drivers/misc using the same style of API as the grove drivers I mentioned earlier.